File: test/java/brs/web/api/http/handler/EscrowSignTest.java
package brs.web.api.http.handler;

import brs.*;
import brs.Escrow.DecisionType;
import brs.common.QuickMocker;
import brs.common.QuickMocker.MockParam;
import brs.fluxcapacitor.FluxCapacitor;
import brs.fluxcapacitor.FluxValues;
import brs.services.EscrowService;
import brs.services.ParameterService;
import brs.util.JSON;
import brs.web.api.http.common.APITransactionManager;
import com.google.gson.JsonObject;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

import javax.servlet.http.HttpServletRequest;

import static brs.TransactionType.AdvancedPayment.ESCROW_SIGN;
import static brs.web.api.http.common.Parameters.DECISION_PARAMETER;
import static brs.web.api.http.common.Parameters.ESCROW_PARAMETER;
import static brs.web.api.http.common.ResultFields.ERROR_CODE_RESPONSE;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.when;
import static org.powermock.api.mockito.PowerMockito.mock;
import static org.powermock.api.mockito.PowerMockito.mockStatic;

@RunWith(PowerMockRunner.class)
@PrepareForTest(Signum.class)
public class EscrowSignTest extends AbstractTransactionTest {

  private ParameterService parameterServiceMock;
  private Blockchain blockchainMock;
  private EscrowService escrowServiceMock;
  private APITransactionManager apiTransactionManagerMock;

  private EscrowSign t;

  @Before
  public void setUp() {
    parameterServiceMock = mock(ParameterService.class);
    blockchainMock = mock(Blockchain.class);
    escrowServiceMock = mock(EscrowService.class);
    apiTransactionManagerMock = mock(APITransactionManager.class);

    t = new EscrowSign(parameterServiceMock, blockchainMock, escrowServiceMock, apiTransactionManagerMock);
  }

  @Test
  public void processRequest_positiveAsEscrowSender() throws SignumException {
    final long escrowId = 5;
    final long senderId = 6;

    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(ESCROW_PARAMETER, escrowId),
        new MockParam(DECISION_PARAMETER, "release")
    );

    final Escrow escrow = mock(Escrow.class);
    when(escrow.getSenderId()).thenReturn(senderId);
    when(escrow.getRecipientId()).thenReturn(2L);

    final Account sender = mock(Account.class);
    when(sender.getId()).thenReturn(senderId);

    mockStatic(Signum.class);
    final FluxCapacitor fluxCapacitor = QuickMocker.fluxCapacitorEnabledFunctionalities(FluxValues.DIGITAL_GOODS_STORE);
    when(Signum.getFluxCapacitor()).thenReturn(fluxCapacitor);
    doReturn(Constants.FEE_QUANT_SIP3).when(fluxCapacitor).getValue(eq(FluxValues.FEE_QUANT));

    when(escrowServiceMock.getEscrowTransaction(eq(escrowId))).thenReturn(escrow);
    when(parameterServiceMock.getSenderAccount(eq(req))).thenReturn(sender);

    final Attachment.AdvancedPaymentEscrowSign attachment = (brs.Attachment.AdvancedPaymentEscrowSign) attachmentCreatedTransaction(() -> t.processRequest(req),
        apiTransactionManagerMock);
    assertNotNull(attachment);

    assertEquals(ESCROW_SIGN, attachment.getTransactionType());
    assertEquals(DecisionType.RELEASE, attachment.getDecision());
  }

  @Test
  public void processRequest_positiveAsEscrowRecipient() throws SignumException {
    final long escrowId = 5;
    final long senderId = 6;

    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(ESCROW_PARAMETER, escrowId),
        new MockParam(DECISION_PARAMETER, "refund")
    );

    final Escrow escrow = mock(Escrow.class);
    when(escrow.getSenderId()).thenReturn(1L);
    when(escrow.getRecipientId()).thenReturn(senderId);

    final Account sender = mock(Account.class);
    when(sender.getId()).thenReturn(senderId);

    when(escrowServiceMock.getEscrowTransaction(eq(escrowId))).thenReturn(escrow);
    when(parameterServiceMock.getSenderAccount(eq(req))).thenReturn(sender);

    mockStatic(Signum.class);
    final FluxCapacitor fluxCapacitor = QuickMocker.fluxCapacitorEnabledFunctionalities(FluxValues.DIGITAL_GOODS_STORE);
    when(Signum.getFluxCapacitor()).thenReturn(fluxCapacitor);
    doReturn(Constants.FEE_QUANT_SIP3).when(fluxCapacitor).getValue(eq(FluxValues.FEE_QUANT));

    final Attachment.AdvancedPaymentEscrowSign attachment = (brs.Attachment.AdvancedPaymentEscrowSign) attachmentCreatedTransaction(() -> t.processRequest(req),
        apiTransactionManagerMock);
    assertNotNull(attachment);

    assertEquals(ESCROW_SIGN, attachment.getTransactionType());
    assertEquals(DecisionType.REFUND, attachment.getDecision());
  }

  @Test
  public void processRequest_positiveAsEscrowSigner() throws SignumException {
    final long escrowId = 5;
    final long senderId = 6;

    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(ESCROW_PARAMETER, escrowId),
        new MockParam(DECISION_PARAMETER, "refund")
    );

    final Escrow escrow = mock(Escrow.class);
    when(escrow.getRecipientId()).thenReturn(1L);
    when(escrow.getSenderId()).thenReturn(2L);

    final Account sender = mock(Account.class);
    when(sender.getId()).thenReturn(senderId);

    mockStatic(Signum.class);
    final FluxCapacitor fluxCapacitor = QuickMocker.fluxCapacitorEnabledFunctionalities(FluxValues.DIGITAL_GOODS_STORE);
    when(Signum.getFluxCapacitor()).thenReturn(fluxCapacitor);
    doReturn(Constants.FEE_QUANT_SIP3).when(fluxCapacitor).getValue(eq(FluxValues.FEE_QUANT));

    when(escrowServiceMock.isIdSigner(eq(senderId), eq(escrow))).thenReturn(true);

    when(escrowServiceMock.getEscrowTransaction(eq(escrowId))).thenReturn(escrow);
    when(parameterServiceMock.getSenderAccount(eq(req))).thenReturn(sender);

    final Attachment.AdvancedPaymentEscrowSign attachment = (brs.Attachment.AdvancedPaymentEscrowSign) attachmentCreatedTransaction(() -> t.processRequest(req),
        apiTransactionManagerMock);
    assertNotNull(attachment);

    assertEquals(ESCROW_SIGN, attachment.getTransactionType());
    assertEquals(DecisionType.REFUND, attachment.getDecision());
  }

  @Test
  public void processRequest_invalidEscrowId() throws SignumException {
    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(ESCROW_PARAMETER, "NotANumber")
    );

    final JsonObject result = (JsonObject) t.processRequest(req);

    assertEquals(3, JSON.getAsInt(result.get(ERROR_CODE_RESPONSE)));
  }

  @Test
  public void processRequest_escrowNotFound() throws SignumException {
    final long escrowId = 5;

    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(ESCROW_PARAMETER, escrowId)
    );

    when(escrowServiceMock.getEscrowTransaction(eq(escrowId))).thenReturn(null);

    final JsonObject result = (JsonObject) t.processRequest(req);

    assertEquals(5, JSON.getAsInt(result.get(ERROR_CODE_RESPONSE)));
  }

  @Test
  public void processRequest_invalidDecisionType() throws SignumException {
    final long escrowId = 5;

    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(ESCROW_PARAMETER, escrowId),
        new MockParam(DECISION_PARAMETER, "notADecisionValue")
    );

    final Escrow escrow = mock(Escrow.class);

    when(escrowServiceMock.getEscrowTransaction(eq(escrowId))).thenReturn(escrow);

    final JsonObject result = (JsonObject) t.processRequest(req);

    assertEquals(5, JSON.getAsInt(result.get(ERROR_CODE_RESPONSE)));
  }

  @Test
  public void processRequest_invalidSender() throws SignumException {
    final long escrowId = 5;
    final long senderId = 6;

    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(ESCROW_PARAMETER, escrowId),
        new MockParam(DECISION_PARAMETER, "refund")
    );

    final Escrow escrow = mock(Escrow.class);
    when(escrow.getSenderId()).thenReturn(1L);
    when(escrow.getRecipientId()).thenReturn(2L);

    when(escrowServiceMock.isIdSigner(eq(senderId), eq(escrow))).thenReturn(false);

    final Account sender = mock(Account.class);
    when(sender.getId()).thenReturn(senderId);

    when(escrowServiceMock.getEscrowTransaction(eq(escrowId))).thenReturn(escrow);
    when(parameterServiceMock.getSenderAccount(eq(req))).thenReturn(sender);

    final JsonObject result = (JsonObject) t.processRequest(req);

    assertEquals(5, JSON.getAsInt(result.get(ERROR_CODE_RESPONSE)));
  }

  @Test
  public void processRequest_senderCanOnlyRelease() throws SignumException {
    final long escrowId = 5;
    final long senderId = 6;

    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(ESCROW_PARAMETER, escrowId),
        new MockParam(DECISION_PARAMETER, "refund")
    );

    final Escrow escrow = mock(Escrow.class);
    when(escrow.getSenderId()).thenReturn(senderId);

    final Account sender = mock(Account.class);
    when(sender.getId()).thenReturn(senderId);

    when(escrowServiceMock.getEscrowTransaction(eq(escrowId))).thenReturn(escrow);
    when(parameterServiceMock.getSenderAccount(eq(req))).thenReturn(sender);

    final JsonObject result = (JsonObject) t.processRequest(req);

    assertEquals(4, JSON.getAsInt(result.get(ERROR_CODE_RESPONSE)));
  }

  @Test
  public void processRequest_recipientCanOnlyRefund() throws SignumException {
    final long escrowId = 5;
    final long senderId = 6;

    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(ESCROW_PARAMETER, escrowId),
        new MockParam(DECISION_PARAMETER, "release")
    );

    final Escrow escrow = mock(Escrow.class);
    when(escrow.getRecipientId()).thenReturn(senderId);

    final Account sender = mock(Account.class);
    when(sender.getId()).thenReturn(senderId);

    when(escrowServiceMock.getEscrowTransaction(eq(escrowId))).thenReturn(escrow);
    when(parameterServiceMock.getSenderAccount(eq(req))).thenReturn(sender);

    final JsonObject result = (JsonObject) t.processRequest(req);

    assertEquals(4, JSON.getAsInt(result.get(ERROR_CODE_RESPONSE)));
  }
}

File: src/test/java/brs/http/EscrowSignTest.kt
package brs.http

import brs.*
import brs.Escrow.DecisionType
import brs.common.QuickMocker
import brs.common.QuickMocker.MockParam
import brs.fluxcapacitor.FluxValues
import brs.http.common.Parameters.DECISION_PARAMETER
import brs.http.common.Parameters.ESCROW_PARAMETER
import brs.http.common.ResultFields.ERROR_CODE_RESPONSE
import brs.services.EscrowService
import brs.services.ParameterService
import brs.transaction.TransactionType
import brs.util.JSON
import com.google.gson.JsonObject
import com.nhaarman.mockitokotlin2.doReturn
import com.nhaarman.mockitokotlin2.eq
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import kotlinx.coroutines.runBlocking
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import javax.servlet.http.HttpServletRequest

@RunWith(JUnit4::class)
class EscrowSignTest : AbstractTransactionTest() {

    private lateinit var parameterServiceMock: ParameterService
    private lateinit var blockchainMock: Blockchain
    private lateinit var escrowServiceMock: EscrowService
    private lateinit var apiTransactionManagerMock: APITransactionManager
    private lateinit var dp: DependencyProvider
    private lateinit var t: EscrowSign

    @Before
    fun setUp() {
        parameterServiceMock = mock<ParameterService>()
        blockchainMock = mock<Blockchain>()
        escrowServiceMock = mock<EscrowService>()
        apiTransactionManagerMock = mock<APITransactionManager>()
        dp = QuickMocker.dependencyProvider(parameterServiceMock!!, blockchainMock!!, escrowServiceMock!!, apiTransactionManagerMock!!)
        t = EscrowSign(dp)
    }

    @Test
    fun processRequest_positiveAsEscrowSender() = runBlocking {
        val escrowId: Long = 5
        val senderId: Long = 6

        val request = QuickMocker.httpServletRequest(
                MockParam(ESCROW_PARAMETER, escrowId),
                MockParam(DECISION_PARAMETER, "release")
        )

        val escrow = mock<Escrow>()
        whenever(escrow.senderId).doReturn(senderId)
        whenever(escrow.recipientId).doReturn(2L)

        val sender = mock<Account>()
        whenever(sender.id).doReturn(senderId)
        dp.fluxCapacitor = QuickMocker.fluxCapacitorEnabledFunctionalities(FluxValues.DIGITAL_GOODS_STORE)
        dp.transactionTypes = TransactionType.getTransactionTypes(dp)

        whenever(escrowServiceMock!!.getEscrowTransaction(eq(escrowId))).doReturn(escrow)
        whenever(parameterServiceMock!!.getSenderAccount(eq<HttpServletRequest>(request))).doReturn(sender)

        val attachment = attachmentCreatedTransaction({ t!!.processRequest(request) },
                apiTransactionManagerMock!!) as Attachment.AdvancedPaymentEscrowSign
        assertNotNull(attachment)

        assertTrue(attachment.transactionType is brs.transaction.advancedPayment.EscrowSign)
        assertEquals(DecisionType.RELEASE, attachment.decision)
    }

    @Test
    fun processRequest_positiveAsEscrowRecipient() = runBlocking {
        val escrowId: Long = 5
        val senderId: Long = 6

        val request = QuickMocker.httpServletRequest(
                MockParam(ESCROW_PARAMETER, escrowId),
                MockParam(DECISION_PARAMETER, "refund")
        )

        val escrow = mock<Escrow>()
        whenever(escrow.senderId).doReturn(1L)
        whenever(escrow.recipientId).doReturn(senderId)

        val sender = mock<Account>()
        whenever(sender.id).doReturn(senderId)

        whenever(escrowServiceMock!!.getEscrowTransaction(eq(escrowId))).doReturn(escrow)
        whenever(parameterServiceMock!!.getSenderAccount(eq<HttpServletRequest>(request))).doReturn(sender)
        dp.fluxCapacitor = QuickMocker.fluxCapacitorEnabledFunctionalities(FluxValues.DIGITAL_GOODS_STORE)
        dp.transactionTypes = TransactionType.getTransactionTypes(dp)

        val attachment = attachmentCreatedTransaction({ t!!.processRequest(request) },
                apiTransactionManagerMock!!) as Attachment.AdvancedPaymentEscrowSign
        assertNotNull(attachment)

        assertTrue(attachment.transactionType is brs.transaction.advancedPayment.EscrowSign)
        assertEquals(DecisionType.REFUND, attachment.decision)
    }

    @Test
    fun processRequest_positiveAsEscrowSigner() = runBlocking {
        val escrowId: Long = 5
        val senderId: Long = 6

        val request = QuickMocker.httpServletRequest(
                MockParam(ESCROW_PARAMETER, escrowId),
                MockParam(DECISION_PARAMETER, "refund")
        )

        val escrow = mock<Escrow>()
        whenever(escrow.recipientId).doReturn(1L)
        whenever(escrow.senderId).doReturn(2L)

        val sender = mock<Account>()
        whenever(sender.id).doReturn(senderId)
        dp.fluxCapacitor = QuickMocker.fluxCapacitorEnabledFunctionalities(FluxValues.DIGITAL_GOODS_STORE)
        dp.transactionTypes = TransactionType.getTransactionTypes(dp)

        whenever(escrowServiceMock!!.isIdSigner(eq(senderId), eq(escrow))).doReturn(true)

        whenever(escrowServiceMock!!.getEscrowTransaction(eq(escrowId))).doReturn(escrow)
        whenever(parameterServiceMock!!.getSenderAccount(eq<HttpServletRequest>(request))).doReturn(sender)

        val attachment = attachmentCreatedTransaction({ t!!.processRequest(request) },
                apiTransactionManagerMock!!) as Attachment.AdvancedPaymentEscrowSign
        assertNotNull(attachment)

        assertTrue(attachment.transactionType is brs.transaction.advancedPayment.EscrowSign)
        assertEquals(DecisionType.REFUND, attachment.decision)
    }

    @Test
    fun processRequest_invalidEscrowId() = runBlocking {
        val request = QuickMocker.httpServletRequest(
                MockParam(ESCROW_PARAMETER, "NotANumber")
        )

        val result = t!!.processRequest(request) as JsonObject

        assertEquals(3, JSON.getAsInt(result.get(ERROR_CODE_RESPONSE)).toLong())
    }

    @Test
    fun processRequest_escrowNotFound() = runBlocking {
        val escrowId: Long = 5

        val request = QuickMocker.httpServletRequest(
                MockParam(ESCROW_PARAMETER, escrowId)
        )

        whenever(escrowServiceMock!!.getEscrowTransaction(eq(escrowId))).doReturn(null)

        val result = t!!.processRequest(request) as JsonObject

        assertEquals(5, JSON.getAsInt(result.get(ERROR_CODE_RESPONSE)).toLong())
    }

    @Test
    fun processRequest_invalidDecisionType() = runBlocking {
        val escrowId: Long = 5

        val request = QuickMocker.httpServletRequest(
                MockParam(ESCROW_PARAMETER, escrowId),
                MockParam(DECISION_PARAMETER, "notADecisionValue")
        )

        val escrow = mock<Escrow>()

        whenever(escrowServiceMock!!.getEscrowTransaction(eq(escrowId))).doReturn(escrow)

        val result = t!!.processRequest(request) as JsonObject

        assertEquals(5, JSON.getAsInt(result.get(ERROR_CODE_RESPONSE)).toLong())
    }

    @Test
    fun processRequest_invalidSender() = runBlocking {
        val escrowId: Long = 5
        val senderId: Long = 6

        val request = QuickMocker.httpServletRequest(
                MockParam(ESCROW_PARAMETER, escrowId),
                MockParam(DECISION_PARAMETER, "refund")
        )

        val escrow = mock<Escrow>()
        whenever(escrow.senderId).doReturn(1L)
        whenever(escrow.recipientId).doReturn(2L)

        whenever(escrowServiceMock!!.isIdSigner(eq(senderId), eq(escrow))).doReturn(false)

        val sender = mock<Account>()
        whenever(sender.id).doReturn(senderId)

        whenever(escrowServiceMock!!.getEscrowTransaction(eq(escrowId))).doReturn(escrow)
        whenever(parameterServiceMock!!.getSenderAccount(eq<HttpServletRequest>(request))).doReturn(sender)

        val result = t!!.processRequest(request) as JsonObject

        assertEquals(5, JSON.getAsInt(result.get(ERROR_CODE_RESPONSE)).toLong())
    }

    @Test
    fun processRequest_senderCanOnlyRelease() = runBlocking {
        val escrowId: Long = 5
        val senderId: Long = 6

        val request = QuickMocker.httpServletRequest(
                MockParam(ESCROW_PARAMETER, escrowId),
                MockParam(DECISION_PARAMETER, "refund")
        )

        val escrow = mock<Escrow>()
        whenever(escrow.senderId).doReturn(senderId)

        val sender = mock<Account>()
        whenever(sender.id).doReturn(senderId)

        whenever(escrowServiceMock!!.getEscrowTransaction(eq(escrowId))).doReturn(escrow)
        whenever(parameterServiceMock!!.getSenderAccount(eq<HttpServletRequest>(request))).doReturn(sender)

        val result = t!!.processRequest(request) as JsonObject

        assertEquals(4, JSON.getAsInt(result.get(ERROR_CODE_RESPONSE)).toLong())
    }

    @Test
    fun processRequest_recipientCanOnlyRefund() = runBlocking {
        val escrowId: Long = 5
        val senderId: Long = 6

        val request = QuickMocker.httpServletRequest(
                MockParam(ESCROW_PARAMETER, escrowId),
                MockParam(DECISION_PARAMETER, "release")
        )

        val escrow = mock<Escrow>()
        whenever(escrow.recipientId).doReturn(senderId)

        val sender = mock<Account>()
        whenever(sender.id).doReturn(senderId)

        whenever(escrowServiceMock!!.getEscrowTransaction(eq(escrowId))).doReturn(escrow)
        whenever(parameterServiceMock!!.getSenderAccount(eq<HttpServletRequest>(request))).doReturn(sender)

        val result = t!!.processRequest(request) as JsonObject

        assertEquals(4, JSON.getAsInt(result.get(ERROR_CODE_RESPONSE)).toLong())
    }
}
